File: services/gpu/common/las_common/webhook.py
from __future__ import annotations

from typing import Optional

from .config import settings


class ProgressReporter:
    """Posts JobProgressWebhook events back to the control plane.

    Authenticated with the shared INTERNAL_SERVICE_TOKEN (no user auth yet).
    Matches packages/protocol JobProgressWebhook.
    """

    def __init__(self, job_id: str) -> None:
        self.job_id = job_id

    def report(
        self,
        status: str,
        progress: Optional[float] = None,
        message: Optional[str] = None,
        output_key: Optional[str] = None,
        error: Optional[str] = None,
    ) -> None:
        import httpx  # lazy

        body = {"jobId": self.job_id, "status": status}
        if progress is not None:
            body["progress"] = progress
        if message is not None:
            body["message"] = message
        if output_key is not None:
            body["outputKey"] = output_key
        if error is not None:
            body["error"] = error

        try:
            httpx.post(
                f"{settings.control_api_url}/api/internal/jobs/progress",
                json=body,
                headers={"authorization": f"Bearer {settings.internal_token}"},
                timeout=10.0,
            )
        except Exception:
            # Progress reporting is best-effort; never crash the pipeline on it.
            pass

File: services/engine-three/src/webhook.ts
import type { Config } from './config.js';

export class ProgressReporter {
  constructor(
    private readonly jobId: string,
    private readonly cfg: Config,
  ) {}

  async report(opts: {
    status: string;
    progress?: number;
    message?: string;
    outputKey?: string;
    error?: string;
  }): Promise<void> {
    const body: Record<string, unknown> = { jobId: this.jobId, status: opts.status };
    if (opts.progress !== undefined) body.progress = opts.progress;
    if (opts.message) body.message = opts.message;
    if (opts.outputKey) body.outputKey = opts.outputKey;
    if (opts.error) body.error = opts.error;

    try {
      await fetch(`${this.cfg.controlApiUrl}/api/internal/jobs/progress`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          authorization: `Bearer ${this.cfg.internalToken}`,
        },
        body: JSON.stringify(body),
      });
    } catch {
      // Best-effort — never fail the render on webhook errors.
    }
  }
}
